projects/com.oracle.truffle.llvm.writer/src/com/oracle/truffle/llvm/writer/facades/FunctionV32Writer.java

```diff
@@ -0,0 +1,221 @@
+            InstructionToLLVMIRVisitor instructionVisitor = new InstructionToLLVMIRVisitor(sb, labels);
+
+            @Override
+            public void visit(InstructionBlock block) {
+                sb.append(createBranchLabel(block) + ":\n");
+                block.accept(instructionVisitor);
+            }
+        });
+
+        sb.append("}");
+        return sb.toString();
+    }
+
+    public static String createFunctionHeader(FunctionDefinition function, Map<Symbol, String> labels) {
+        List<String> sb = new ArrayList<>();
+        sb.add("define");
+        sb.add(function.getReturnType().toString());
+
+        List<String> args = new ArrayList<>();
+        for (FunctionParameter fArg : function.getParameters()) {
+            args.add(fArg.getType() + " " + addLabelToSymbolsMap(labels, fArg));
+        }
+        if (function.isVarArg()) {
+            args.add("...");
+        }
+
+        sb.add(function.getName() + "(" + args.stream().collect(Collectors.joining(", ")) + ")");
+        return sb.stream().collect(Collectors.joining(" "));
+    }
+
+    private static String addLabelToSymbolsMap(Map<Symbol, String> labels, Symbol curSymbol) {
+        // we simply use named variables, so we don't have to worry about enumeration
+        String returnVar = "%sym_" + (labels.size() + 1);
+        labels.put(curSymbol, returnVar);
+        return returnVar;
+    }
+
+    private static String createBranchLabel(InstructionBlock block) {
+        // we simply use named blocks, so we don't have to worry about enumeration
+        return "block_" + block.getBlockIndex();
+    }
+
+    public static class InstructionToLLVMIRVisitor implements InstructionVisitorAdapter {
+        private final StringBuilder llvmir;
+        final Map<Symbol, String> labels;
+
+        public InstructionToLLVMIRVisitor(StringBuilder llvmir, Map<Symbol, String> labels) {
+            this.llvmir = llvmir;
+            this.labels = labels;
+        }
+
+        private void addInstructionString(String str) {
+            llvmir.append("    "); // spaces at start
+            llvmir.append(str);
+            llvmir.append("\n");
+        }
+
+        private String newLabel(Symbol curInstr) {
+            return addLabelToSymbolsMap(labels, curInstr);
+        }
+
+        private String createSymbolLabel(Symbol s) {
+            return s.getType() + " " + createSymbolLabelWithoutType(s);
+        }
+
+        private String createSymbolLabelWithoutType(Symbol s) {
+            if (s instanceof Constant) {
+                return s.toString(); // TODO: needs check
+            } else {
+                return labels.get(s);
+            }
+        }
+
+        private static String createSymbolLabel(InstructionBlock s) {
+            return "label %" + createBranchLabel(s);
+        }
+
+        @Override
+        public void visit(AllocateInstruction allocate) {
+            List<String> sb = new ArrayList<>();
```

Comment:
Could you please use a StringBuilder or, better yet, String.format() instead of lists in these methods.